odoo/addons/test_read_group/tests/test_m2m_grouping.py

```diff
@@ -0,0 +1,193 @@
+        # expected values
+        expected_user_by_tasks = [
+            {
+                'task_ids_count': 2,
+                'name': users.mapped('name'),
+                '__domain': [('task_ids', 'in', (tasks[0].id,))],
+                'task_ids': tasks[0].id,
+            },
+            {
+                'task_ids_count': 2,
+                'name': users.mapped('name'),
+                '__domain': [('task_ids', 'in', (tasks[1].id,))],
+                'task_ids': tasks[1].id,
+            },
+        ]
+        expected_task_by_users = [
+            {
+                'user_ids_count': 2,
+                'name': tasks.mapped('name'),
+                '__domain': [('user_ids', 'in', (users[0].id,))],
+                'user_ids': users[0].id,
+            },
+            {
+                'user_ids_count': 2,
+                'name': tasks.mapped('name'),
+                '__domain': [('user_ids', 'in', (users[1].id,))],
+                'user_ids': users[1].id,
+            },
+        ]
+
+        self.assertEqual(user_by_tasks, expected_user_by_tasks)
+        self.assertEqual(task_by_users, expected_task_by_users)
+
+    def test_comodel_ir_rules(self):
+        users_model_id = self.env['ir.model']._get(self.ResUsers._name)
+        self.env['ir.rule'].create({
+            'name': "Only The Lone Wanderer allowed",
+            'model_id': users_model_id.id,
+            'domain_force': [('name', '=', 'The Lone Wanderer')],
+        })
+        users = self.ResUsers.create([{'name': 'The Lone Wanderer'}, {'name': 'Yes man'}])
+        task = self.ProjectTask.create({
+            'name': "Find Yes Man",
+            'user_ids': [fields.Command.set(users.ids)],
+        })
+
+        # ir.rule should not apply
+        as_admin = task.read_group(
+            domain=[],
+            fields=['name:array_agg'],
+            groupby=['user_ids'],
+        )
+        self.assertEqual(len(as_admin), 2)
+        fetched_users = self.ResUsers.browse([r['user_ids'] for r in as_admin])
+        self.assertEqual(fetched_users.mapped('name'), ['The Lone Wanderer', 'Yes man'])
+
+        # ir.rule should apply
+        task = task.with_user(self.browse_ref('base.user_demo'))
+        expected = """
+        SELECT
+            min("test_read_group_project_task".id) AS id,
+            count("test_read_group_project_task".id) AS "user_ids_count",
+            array_agg("test_read_group_project_task"."name") AS "name",
+            "test_read_group_project_task__id__test_read_group_res_users_id"."id" as "user_ids"
+        FROM "test_read_group_project_task"
+        LEFT JOIN "test_read_group_project_task_res_users_rel"
+            AS "test_read_group_project_task__id"
+            ON ("test_read_group_project_task"."id"
+                = "test_read_group_project_task__id"."test_read_group_project_task_id")
+        LEFT JOIN "test_read_group_res_users"
+            AS "test_read_group_project_task__id__test_read_group_res_users_id"
```

Comment:
The aliases used by both joins are a bit confusing, and may lead to errors, since they are supposed to be unique.
The convention is to combine the model alias with the name of the field used for the join.
Both aliases should somehow follow the convention *model__field*, something like
* alias of the relation table: *model__relation* (`test_read_group_project_task__test_read_group_project_task_res_users_rel`)
* alias of the comodel table: *model__field* (`test_read_group_project_task__user_ids`)